File: app/src/main/java/com/learning/helloworld/StudentsAdapter.java
package com.learning.helloworld;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

class StudentViewHolder extends RecyclerView.ViewHolder {
    private TextView nameTextView;
    private Student student;
    private StudentsFragment.StudentClickListener studentClickListener;

    public StudentViewHolder(@NonNull View itemView, StudentsFragment.StudentClickListener studentClickListener) {
        super(itemView);
        nameTextView = itemView.findViewById(R.id.name);
        this.studentClickListener = studentClickListener;
    }

    void bind(Student student) {
        this.student = student;
        nameTextView.setText(student.firstName + " " + student.lastName);
        nameTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                studentClickListener.onStudentClicked(StudentViewHolder.this.student);
            }
        });
    }
}

public class StudentsAdapter extends RecyclerView.Adapter<StudentViewHolder> {
    List<Student> students;
    private StudentsFragment.StudentClickListener studentClickListener;

    StudentsAdapter(List<Student> students, StudentsFragment.StudentClickListener studentClickListener) {
        this.students = students;
        this.studentClickListener = studentClickListener;
    }

    @NonNull
    @Override
    public StudentViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = (LayoutInflater) parent.getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View itemView = inflater.inflate(R.layout.student, parent, false);
        return new StudentViewHolder(itemView, studentClickListener);
    }

    @Override
    public void onBindViewHolder(@NonNull StudentViewHolder holder, int position) {
        holder.bind(students.get(position));
    }

    @Override
    public int getItemCount() {
        return students.size();
    }
}

File: app/src/main/java/com/learning/helloworld/StudentsAdapter.kt
package com.learning.helloworld

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView

class StudentViewHolder(
    itemView: ConstraintLayout,
    val onClick : (Student) -> Unit
) : RecyclerView.ViewHolder(itemView) {
    private var nameTextView : TextView? = itemView.findViewById(R.id.name)

    fun bind(student: Student) {
        nameTextView?.text = "${student.firstName} ${student.lastName}"
        nameTextView?.setOnClickListener {
            onClick(student)
        }
    }
}

class StudentsAdapter(
    private val students: List<Student>,
    private val onClick : (Student) -> Unit
) : RecyclerView.Adapter<StudentViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): StudentViewHolder {
        val inflater = parent.context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val itemView = inflater.inflate(R.layout.student, parent, false) as ConstraintLayout
        return StudentViewHolder(itemView, onClick)
    }

    override fun getItemCount(): Int {
        return students.size
    }

    override fun onBindViewHolder(holder: StudentViewHolder, position: Int) {
        holder.bind(students[position])
    }

}